source/solvers/navier_stokes_base.cc

```diff
@@ -1987,24 +1968,41 @@ NavierStokesBase<dim, VectorType, DofsType>::define_zero_constraints()
 
 template <int dim, typename VectorType, typename DofsType>
 void
-NavierStokesBase<dim, VectorType, DofsType>::update_mortar_coupling()
+NavierStokesBase<dim, VectorType, DofsType>::update_mortar_configuration()
 {
-  TimerOutput::Scope t(this->computing_timer, "Update mortar");
+  if (!this->simulation_parameters.mortar_parameters.enable)
+    return;
+
+  bool refinement_step;
+  if (this->simulation_parameters.mesh_adaptation.refinement_at_frequency)
+    refinement_step = this->simulation_control->get_step_number() %
+                        this->simulation_parameters.mesh_adaptation.frequency ==
+                      0;
+  else
+    refinement_step = this->simulation_control->get_step_number() == 0;
 
-  if (!this->simulation_parameters.mortar.enable)
+  if (this->simulation_control->is_at_start() || !refinement_step ||
+      this->simulation_parameters.mesh_adaptation.type ==
+        Parameters::MeshAdaptation::Type::none)
+    setup_dofs();
```

Comment:
I would add a comment here to explain what you are doing. There's a bunch of if, they make sense, but still it's not trivial to understand.